Quantifiers assume that ranges are always usizes

<!--
If this is a security issue, please report it following the
[security reporting procedure](https://github.com/model-checking/kani/security/policy).
-->

<!--
Thank you for filing a bug report! 🐛
Please provide a short summary of the issue, along with the information necessary to replicate.
-->

This is a tracking issue for fixing the [known quantifiers limitation](https://model-checking.github.io/kani/reference/experimental/quantifiers.html#types-of-quantified-variables) that the ranges must be usizes.

I tried this code:
<!--
If the reproducing code is small, please post it here.
Otherwise, please attach the relevant files to this issue.
Alternatively, post a link to the repository and branch that exposes the issue.
-->
```rust
#[kani::requires(min != 0 && max != 0)]
#[kani::ensures(|result| *result != 0 && max % *result == 0 && min % *result == 0)]
#[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
#[kani::recursion]
fn gcd(mut max: u8, mut min: u8) -> u8 {
    if min > max {
        std::mem::swap(&mut max, &mut min);
    }

    let rest = max % min;
    if rest == 0 { min } else { gcd(max, rest) }
}

#[kani::proof_for_contract(gcd)]
fn check_gcd() {
    let max: u8 = kani::any();
    let min: u8 = kani::any();
    gcd(max, min);
}
```

using the following command line invocation:

```
cargo kani -Z function-contracts
```

with Kani version: 0.62.0

I expected to see this happen: verification success

<!--
If Kani crashed, please include a backtrace in the code block by setting
`RUST_BACKTRACE=1` in your environment. E.g. `RUST_BACKTRACE=1 kani ...`
-->
Instead, this happened: 
```
cmzech@80a9971b5e20 playground % cargo kani -Z function-contracts
Kani Rust Verifier 0.62.0 (cargo plugin)
   Compiling playground v0.1.0 (/Users/cmzech/playground)
error[E0308]: mismatched types
 --> src/lib.rs:3:49
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                          -----------------------^^^^^^^---------------------------------------------------
  |                          |                      |
  |                          |                      expected `usize`, found `u8`
  |                          expected due to this
  |
help: you can convert a `u8` to a `usize`
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX.into())| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                                                        +++++++

error[E0631]: type mismatch in closure arguments
  --> src/lib.rs:3:26
   |
3  | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          expected due to this
   |                          found signature defined here
   |                          required by a bound introduced by this call
   |
   = note: expected closure signature `fn(usize) -> _`
              found closure signature `fn(u8) -> _`
note: required by a bound in `kani::internal::kani_forall`
  --> /Users/cmzech/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `kani_forall`
   = note: this error originates in the macro `kani::forall` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider wrapping the function in a closure
  --> /Users/cmzech/kani/library/kani_core/src/lib.rs:203:71
   |
203|                 kani::internal::kani_forall(lower_bound, upper_bound, |arg0: usize| predicate(/* u8 */))
   |                                                                       +++++++++++++          ++++++++++

error[E0308]: mismatched types
 --> src/lib.rs:3:49
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                          -----------------------^^^^^^^---------------------------------------------------
  |                          |                      |
  |                          |                      expected `usize`, found `u8`
  |                          expected due to this
  |
help: you can convert a `u8` to a `usize`
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX.into())| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                                                        +++++++

error[E0631]: type mismatch in closure arguments
  --> src/lib.rs:3:26
   |
3  | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          expected due to this
   |                          found signature defined here
   |                          required by a bound introduced by this call
   |
   = note: expected closure signature `fn(usize) -> _`
              found closure signature `fn(u8) -> _`
note: required by a bound in `kani::internal::kani_forall`
  --> /Users/cmzech/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `kani_forall`
   = note: this error originates in the macro `kani::forall` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider wrapping the function in a closure
  --> /Users/cmzech/kani/library/kani_core/src/lib.rs:203:71
   |
203|                 kani::internal::kani_forall(lower_bound, upper_bound, |arg0: usize| predicate(/* u8 */))
   |                                                                       +++++++++++++          ++++++++++

error[E0308]: mismatched types
 --> src/lib.rs:3:49
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                          -----------------------^^^^^^^---------------------------------------------------
  |                          |                      |
  |                          |                      expected `usize`, found `u8`
  |                          expected due to this
  |
help: you can convert a `u8` to a `usize`
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX.into())| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                                                        +++++++

error[E0631]: type mismatch in closure arguments
  --> src/lib.rs:3:26
   |
3  | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          expected due to this
   |                          found signature defined here
   |                          required by a bound introduced by this call
   |
   = note: expected closure signature `fn(usize) -> _`
              found closure signature `fn(u8) -> _`
note: required by a bound in `kani::internal::kani_forall`
  --> /Users/cmzech/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `kani_forall`
   = note: this error originates in the macro `kani::forall` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider wrapping the function in a closure
  --> /Users/cmzech/kani/library/kani_core/src/lib.rs:203:71
   |
203|                 kani::internal::kani_forall(lower_bound, upper_bound, |arg0: usize| predicate(/* u8 */))
   |                                                                       +++++++++++++          ++++++++++

error[E0308]: mismatched types
 --> src/lib.rs:3:49
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                          -----------------------^^^^^^^---------------------------------------------------
  |                          |                      |
  |                          |                      expected `usize`, found `u8`
  |                          expected due to this
  |
help: you can convert a `u8` to a `usize`
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX.into())| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                                                        +++++++

error[E0631]: type mismatch in closure arguments
  --> src/lib.rs:3:26
   |
3  | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          expected due to this
   |                          found signature defined here
   |                          required by a bound introduced by this call
   |
   = note: expected closure signature `fn(usize) -> _`
              found closure signature `fn(u8) -> _`
note: required by a bound in `kani::internal::kani_forall`
  --> /Users/cmzech/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `kani_forall`
   = note: this error originates in the macro `kani::forall` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider wrapping the function in a closure
  --> /Users/cmzech/kani/library/kani_core/src/lib.rs:203:71
   |
203|                 kani::internal::kani_forall(lower_bound, upper_bound, |arg0: usize| predicate(/* u8 */))
   |                                                                       +++++++++++++          ++++++++++

error[E0308]: mismatched types
 --> src/lib.rs:3:49
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                          -----------------------^^^^^^^---------------------------------------------------
  |                          |                      |
  |                          |                      expected `usize`, found `u8`
  |                          expected due to this
  |
help: you can convert a `u8` to a `usize`
  |
3 | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX.into())| !(max % i == 0 && min % i == 0) || *result >= i))]
  |                                                        +++++++

error[E0631]: type mismatch in closure arguments
  --> src/lib.rs:3:26
   |
3  | #[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX)| !(max % i == 0 && min % i == 0) || *result >= i))]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          expected due to this
   |                          found signature defined here
   |                          required by a bound introduced by this call
   |
   = note: expected closure signature `fn(usize) -> _`
              found closure signature `fn(u8) -> _`
note: required by a bound in `kani::internal::kani_forall`
  --> /Users/cmzech/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `kani_forall`
   = note: this error originates in the macro `kani::forall` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider wrapping the function in a closure
  --> /Users/cmzech/kani/library/kani_core/src/lib.rs:203:71
   |
203|                 kani::internal::kani_forall(lower_bound, upper_bound, |arg0: usize| predicate(/* u8 */))
   |                                                                       +++++++++++++          ++++++++++

Some errors have detailed explanations: E0308, E0631.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 10 previous errors
error: Failed to execute cargo (exit status: 101). Found 10 compilation errors.
````

I had to change the quantifier to this:

```rust
#[kani::ensures(|result| kani::forall!(|i in (1,u8::MAX as usize)| !(max as usize % (i as usize) == 0 && min as usize % (i as usize) == 0) || (*result as usize) >= (i as usize)))]

```

This is quite cumbersome. The error message is very confusing (and repetitive), so short of fixing the problem, we can at least try to improve that.

--- a/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs
+++ b/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs
@@ -206,6 +206,11 @@ impl GotocCodegenBackend {
             None
         };
 
+        // Post-pass: inline remaining function calls in quantifier bodies.
+        // build_quantifier_predicate handles checked arithmetic (StatementExpression
+        // flattening, overflow simplification), but user-defined function calls
+        // (e.g., comp(x, y)) require this post-pass because the called functions
+        // may not be in the symbol table when the quantifier hook runs.
         gcx.handle_quantifiers();
 
         // Split ownership of the context so that the majority of fields can be saved to our results,

--- a/kani-compiler/src/codegen_cprover_gotoc/context/goto_ctx.rs
+++ b/kani-compiler/src/codegen_cprover_gotoc/context/goto_ctx.rs
@@ -354,6 +354,7 @@ impl<'tcx, 'r> GotocCtx<'tcx, 'r> {
 impl GotocCtx<'_, '_> {
     /// Inline all function calls in `expr` as pure expressions.
     /// Prefer `inline_as_pure_expr_toplevel` for the public entry point.
+    #[allow(clippy::collapsible_if, clippy::cmp_owned)]
     fn inline_as_pure_expr(&self, expr: &Expr, visited: &mut HashSet<InternedString>) -> Expr {
         match expr.value() {
             ExprValue::FunctionCall { function, arguments } => {
@@ -392,7 +393,35 @@ impl GotocCtx<'_, '_> {
                 .inline_as_pure_expr(array, visited)
                 .index(self.inline_as_pure_expr(index, visited)),
             ExprValue::Member { lhs, field } => {
-                self.inline_as_pure_expr(lhs, visited).member(*field, &self.symbol_table)
+                let inlined_lhs = self.inline_as_pure_expr(lhs, visited);
+                // If the inlined lhs is a Struct literal, extract the field directly
+                // by numeric index (e.g., field "0" → values[0]).
+                if let ExprValue::Struct { values } = inlined_lhs.value() {
+                    if let Ok(idx) = field.to_string().parse::<usize>() {
+                        if idx < values.len() {
+                            return self.inline_as_pure_expr(&values[idx], visited);
+                        }
+                    }
+                }
+                // Member(OverflowResultOp(a,b), "result") → Op(a,b)
+                // Simplifies checked arithmetic to plain arithmetic for quantifier bodies.
+                if *field == "result" {
+                    if let ExprValue::BinOp { op, lhs: a, rhs: b } = inlined_lhs.value() {
+                        use cbmc::goto_program::BinaryOperator::*;
+                        let simple = match op {
+                            OverflowResultPlus => Some(Plus),
+                            OverflowResultMinus => Some(Minus),
+                            OverflowResultMult => Some(Mult),
+                            _ => None,
+                        };
+                        if let Some(sop) = simple {
+                            return self
+                                .inline_as_pure_expr(a, visited)
+                                .binop(sop, self.inline_as_pure_expr(b, visited));
+                        }
+                    }
+                }
+                inlined_lhs.member(*field, &self.symbol_table)
             }
             ExprValue::Dereference(e) => self.inline_as_pure_expr(e, visited).dereference(),
             ExprValue::AddressOf(e) => Expr::address_of(self.inline_as_pure_expr(e, visited)),
@@ -407,14 +436,57 @@ impl GotocCtx<'_, '_> {
                 self.inline_as_pure_expr(domain, visited),
             ),
             ExprValue::StatementExpression { statements, .. } => {
-                // Extract the final expression from the statement block.
-                // This handles checked arithmetic (Assert + Assume + Expression).
-                for stmt in statements.iter().rev() {
-                    if let StmtBody::Expression(e) = stmt.body() {
-                        return self.inline_as_pure_expr(e, visited);
+                // Flatten StatementExpression by collecting intermediate variable
+                // assignments and substituting them into the final expression.
+                // This handles checked arithmetic patterns like:
+                //   { let temp = overflow_result_plus(a, b); (temp.result, temp.overflowed) }
+                let mut assignments: HashMap<InternedString, Expr> = HashMap::new();
+                let mut final_expr = None;
+                for stmt in statements.iter() {
+                    match stmt.body() {
+                        StmtBody::Decl { lhs, value: Some(val) } => {
+                            if let ExprValue::Symbol { identifier } = lhs.value() {
+                                assignments.insert(*identifier, val.clone());
+                            }
+                        }
+                        StmtBody::Expression(e) => {
+                            final_expr = Some(e.clone());
+                        }
+                        _ => {
+                            // Dropping Assert/Assume statements (overflow/div-by-zero
+                            // checks) from quantifier bodies. These cannot be
+                            // represented as pure expressions.
+                            tracing::debug!(
+                                "Dropping checked arithmetic guard in quantifier body. \
+                                 Overflow and division-by-zero checks are not enforced \
+                                 inside quantifier predicates."
+                            );
+                        }
                     }
                 }
-                expr.clone()
+                if let Some(mut expr) = final_expr {
+                    // Resolve intermediate variables.
+                    // Capped at assignments.len() + 1 to guard against cycles.
+                    for _ in 0..=assignments.len() {
+                        let mut changed = false;
+                        for (sym, rhs) in assignments.iter() {
+                            let (new_expr, did_change) = expr.substitute_symbol(sym, rhs);
+                            expr = new_expr;
+                            changed |= did_change;
+                        }
+                        if !changed {
+                            break;
+                        }
+                    }
+                    // After resolution, extract just the "result" field from
+                    // overflow-checked operations. The pattern is:
+                    //   Member(Struct([Member(overflow_op, "result"), ...]), "0")
+                    // Simplify to just the arithmetic result.
+                    let simplified = simplify_overflow_result(&expr);
+                    self.inline_as_pure_expr(&simplified, visited)
+                } else {
+                    expr.clone()
+                }
             }
             _ => expr.clone(),
         }
@@ -1136,5 +1208,80 @@ impl<'tcx> FnAbiOfHelpers<'tcx> for GotocCtx<'tcx, '_> {
                 }
             }
         }
+    }
+}
+
+/// Simplify overflow-checked arithmetic results for use in quantifier bodies.
+/// Replaces patterns like `Member(OverflowResultPlus(a, b), "result")` with `Plus(a, b)`.
+/// This drops the overflow check, which is acceptable inside quantifier bodies
+/// where CBMC requires side-effect-free expressions.
+#[allow(clippy::collapsible_if, clippy::cmp_owned)]
+fn simplify_overflow_result(expr: &Expr) -> Expr {
+    use cbmc::goto_program::BinaryOperator::*;
+    match expr.value() {
+        // Member(Struct([f0, f1, ...]), "N") → simplify(fN)
+        ExprValue::Member { lhs, field } => {
+            let simplified_lhs = simplify_overflow_result(lhs);
+            match simplified_lhs.value() {
+                // If lhs simplified to a Struct, extract the field by index
+                ExprValue::Struct { values } => {
+                    let field_str = field.to_string();
+                    if let Ok(idx) = field_str.parse::<usize>() {
+                        if idx < values.len() {
+                            return simplify_overflow_result(&values[idx]);
+                        }
+                    }
+                    // "result" field from OverflowResult
+                    if field_str == "result" {
+                        if let ExprValue::BinOp { op, lhs: a, rhs: b } = simplified_lhs.value() {
+                            let simple_op = match op {
+                                OverflowResultPlus => Some(Plus),
+                                OverflowResultMinus => Some(Minus),
+                                OverflowResultMult => Some(Mult),
+                                _ => None,
+                            };
+                            if let Some(op) = simple_op {
+                                return simplify_overflow_result(a)
+                                    .binop(op, simplify_overflow_result(b));
+                            }
+                        }
+                    }
+                    expr.clone()
+                }
+                // Member(OverflowResultPlus(a,b), "result") → Plus(a,b)
+                _ => {
+                    if field.to_string() == "result" {
+                        if let ExprValue::BinOp { op, lhs: a, rhs: b } = simplified_lhs.value() {
+                            let simple_op = match op {
+                                OverflowResultPlus => Some(Plus),
+                                OverflowResultMinus => Some(Minus),
+                                OverflowResultMult => Some(Mult),
+                                _ => None,
+                            };
+                            if let Some(op) = simple_op {
+                                return simplify_overflow_result(a)
+                                    .binop(op, simplify_overflow_result(b));
+                            }
+                        }
+                    }
+                    expr.clone()
+                }
+            }
+        }
+        ExprValue::BinOp { op, lhs, rhs } => {
+            simplify_overflow_result(lhs).binop(*op, simplify_overflow_result(rhs))
+        }
+        ExprValue::Typecast(e) => simplify_overflow_result(e).cast_to(expr.typ().clone()),
+        ExprValue::If { c, t, e } => simplify_overflow_result(c)
+            .ternary(simplify_overflow_result(t), simplify_overflow_result(e)),
+        ExprValue::Struct { .. } => {
+            // Struct values are not recursed into here because constructing a
+            // new Struct expression requires a SymbolTable reference. Instead,
+            // overflow simplification happens when individual fields are accessed
+            // via the Member handler in inline_as_pure_expr, which extracts and
+            // simplifies fields on demand.
+            expr.clone()
+        }
+        _ => expr.clone(),
     }
 }

--- a/kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs
+++ b/kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs
@@ -15,7 +15,8 @@ use crate::kani_middle::kani_functions::{KaniFunction, KaniHook, try_get_kani_fu
 use crate::unwrap_or_return_codegen_unimplemented_stmt;
 use cbmc::goto_program::CIntType;
 use cbmc::goto_program::Symbol as GotoSymbol;
-use cbmc::goto_program::{BuiltinFn, Expr, Location, Stmt, Type};
+use cbmc::goto_program::{BuiltinFn, Expr, Location, Stmt, StmtBody, SymbolValues, Type};
+use cbmc::{InternedString, goto_program::ExprValue};
 use rustc_hir::LangItem;
 use rustc_middle::ty::TyCtxt;
 use rustc_public::mir::mono::Instance;
@@ -955,21 +956,11 @@ fn handle_quantifier(
     let target = target.unwrap();
     let lower_bound = &fargs[0];
     let upper_bound = &fargs[1];
-    let closure_call_expr = find_closure_call_expr(&instance, gcx, loc)
-        .unwrap_or_else(|| unreachable!("Failed to find closure call expression"));
-    let closure_arg = fargs[2].clone();
-    let predicate = if closure_arg.is_symbol() {
-        Expr::address_of(closure_arg)
-    } else {
-        let predicate_ty = fargs[2].typ().clone().to_pointer();
-        Expr::nondet(predicate_ty)
-    };
 
-    // Quantified variable.
+    // Create a fresh quantified variable.
     let base_name = "kani_quantified_var".to_string();
     let mut counter = 0;
     let mut unique_name = format!("{base_name}_{counter}");
-    // Ensure the name is not already in the symbol table
     while gcx.symbol_table.lookup(&unique_name).is_some() {
         counter += 1;
         unique_name = format!("{base_name}_{counter}");
@@ -981,23 +972,40 @@ fn handle_quantifier(
         new_symbol.to_expr()
     };
 
+    // Build the predicate as a pure expression by substituting the closure
+    // parameter with the quantified variable. This avoids generating function
+    // calls inside the quantifier body, which CBMC rejects as side effects.
+    // Falls back to a closure function call (resolved by the handle_quantifiers
+    // post-pass) when the predicate contains StatementExpression nodes that
+    // substitute_symbol cannot recurse into.
+    let predicate_expr =
+        match build_quantifier_predicate(gcx, &instance, &fargs[2], &new_variable_expr) {
+            Some(expr) => expr,
+            None => {
+                // Fallback: emit a closure function call for the post-pass to inline.
+                let closure_call_expr = find_closure_call_expr(&instance, gcx, loc)
+                    .unwrap_or_else(|| unreachable!("Failed to find closure call expression"));
+                let predicate = if fargs[2].is_symbol() {
+                    Expr::address_of(fargs[2].clone())
+                } else {
+                    let predicate_ty = fargs[2].typ().clone().to_pointer();
+                    Expr::nondet(predicate_ty)
+                };
+                closure_call_expr.call(vec![predicate, new_variable_expr.clone()])
+            }
+        };
+
     let lower_bound_comparison = lower_bound.clone().le(new_variable_expr.clone());
     let upper_bound_comparison = new_variable_expr.clone().lt(upper_bound.clone());
     let range = lower_bound_comparison.and(upper_bound_comparison);
 
     let quantifier_expr = match quantifier_kind {
         QuantifierKind::ForAll => {
-            let domain = range
-                .clone()
-                .implies(closure_call_expr.call(vec![predicate.clone(), new_variable_expr.clone()]))
-                .and(range.not().implies(Expr::bool_true()));
+            let domain = range.implies(predicate_expr);
             Expr::forall_expr(Type::Bool, new_variable_expr, domain)
         }
         QuantifierKind::Exists => {
-            let domain = range
-                .clone()
-                .and(closure_call_expr.call(vec![predicate.clone(), new_variable_expr.clone()]))
-                .and(range.not().implies(Expr::bool_false()));
+            let domain = range.and(predicate_expr);
             Expr::exists_expr(Type::Bool, new_variable_expr, domain)
         }
     };
@@ -1016,6 +1024,103 @@ fn handle_quantifier(
     )
 }
 
+/// Build a pure expression for the quantifier predicate by looking up the
+/// closure's codegen'd body and substituting its parameter with the quantified
+/// variable. This produces a side-effect-free expression that CBMC can handle
+/// directly, including in nested quantifier contexts.
+///
+/// Returns `None` if the substituted expression still contains side effects
+/// (e.g., `StatementExpression` from checked arithmetic or function calls),
+/// signaling the caller to fall back to the closure function call approach.
+fn build_quantifier_predicate(
+    gcx: &mut GotocCtx,
+    instance: &Instance,
+    closure_arg: &Expr,
+    quantified_var: &Expr,
+) -> Option<Expr> {
+    // Find the closure instance from the generic args.
+    let closure_instance = find_closure_instance(instance)
+        .unwrap_or_else(|| unreachable!("Failed to find closure instance for quantifier"));
+
+    // Look up the closure's codegen'd function body in the symbol table.
+    let closure_name = closure_instance.mangled_name();
+    let closure_body = gcx
+        .symbol_table
+        .lookup(&closure_name)
+        .and_then(|sym| match &sym.value {
+            SymbolValues::Stmt(stmt) => Some(stmt.clone()),
+            _ => None,
+        })
+        .or_else(|| {
+            // If the closure hasn't been codegen'd yet, codegen it now.
+            // SAFETY: We save and restore `current_fn` because `codegen_function`
+            // sets it internally. Without this, the outer function's context would
+            // be lost. This is safe because `codegen_function` only reads/writes
+            // the symbol table and `current_fn` — no other GotocCtx state is
+            // affected by the temporary `None` value during the inner codegen.
+            let saved_fn = gcx.current_fn.take();
+            gcx.codegen_function(closure_instance);
+            gcx.current_fn = saved_fn;
+            gcx.symbol_table.lookup(&closure_name).and_then(|sym| match &sym.value {
+                SymbolValues::Stmt(stmt) => Some(stmt.clone()),
+                _ => None,
+            })
+        })
+        .unwrap_or_else(|| unreachable!("Failed to codegen closure for quantifier"));
+
+    // Get the closure's parameter symbols.
+    let parameters = gcx
+        .symbol_table
+        .lookup_parameters(&closure_name)
+        .unwrap_or_else(|| unreachable!("Failed to find closure parameters"));
+
+    // The closure signature is (closure_env_ref, param) -> bool.
+    // parameters[0] = closure environment reference (&self)
+    // parameters[1] = the quantified variable parameter
+    if parameters.len() != 2 {
+        let msg = format!(
+            "Quantifier closure has {} parameters (expected 2). \
+             The quantifier predicate will evaluate to false.",
+            parameters.len()
+        );
+        gcx.tcx.dcx().warn(msg);
+        return None;
+    }
+    let env_param = parameters[0];
+    let var_param = parameters[1];
+
+    // Extract the return expression from the closure body.
+    let return_expr = extract_return_expr(&closure_body)
+        .unwrap_or_else(|| unreachable!("Failed to extract return expression from closure body"));
+
+    debug!(?env_param, ?var_param, ?return_expr, "quantifier predicate substitution");
+
+    // Build the closure environment expression (address of the closure arg).
+    let env_expr = if closure_arg.is_symbol() {
+        Expr::address_of(closure_arg.clone())
+    } else {
+        let predicate_ty = closure_arg.typ().clone().to_pointer();
+        Expr::nondet(predicate_ty)
+    };
+
+    // Substitute: replace the var parameter with the quantified variable,
+    // and the env parameter with the closure environment.
+    // substitute_symbol does NOT recurse into StatementExpression nodes,
+    // so if the closure body contains checked arithmetic or function calls
+    // that produce StatementExpression, the result will still have side effects.
+    // In that case, return None to signal the caller to use the fallback path.
+    let result = return_expr
+        .substitute_symbol(&var_param, quantified_var)
+        .0
+        .substitute_symbol(&env_param, &env_expr)
+        .0;
+
+    if result.is_side_effect() { None } else { Some(result) }
+}
+
+/// Get the closure's function pointer expression for use in a function call.
+/// This is the fallback path when direct substitution cannot produce a
+/// side-effect-free predicate expression.
 fn find_closure_call_expr(instance: &Instance, gcx: &mut GotocCtx, loc: Location) -> Option<Expr> {
     for arg in instance.args().0.iter() {
         let arg_ty = arg.ty()?;
@@ -1029,6 +1134,106 @@ fn find_closure_call_expr(instance: &Instance, gcx: &mut GotocCtx, loc: Location
         }
     }
     None
+}
+
+/// Find the closure Instance from the quantifier function's generic args.
+fn find_closure_instance(instance: &Instance) -> Option<Instance> {
+    for arg in instance.args().0.iter() {
+        let arg_ty = arg.ty()?;
+        let kind = arg_ty.kind();
+        let arg_kind = kind.rigid()?;
+        if let RigidTy::Closure(def_id, args) = arg_kind {
+            return Instance::resolve_closure(*def_id, args, ClosureKind::Fn).ok();
+        }
+    }
+    None
+}
+
+/// Extract the return expression from a function body, fully resolving all
+/// intermediate variable assignments to produce a self-contained expression
+/// that only references the function parameters.
+fn extract_return_expr(stmt: &Stmt) -> Option<Expr> {
+    // Collect all assignments: symbol -> rhs
+    let mut assignments: HashMap<InternedString, Expr> = HashMap::new();
+    collect_assignments(stmt, &mut assignments);
+
+    // Find the return expression — either a symbol (resolved via assignments)
+    // or a direct expression.
+    let mut expr = match find_return_expr(stmt) {
+        Some(ReturnExpr::Symbol(sym)) => assignments.remove(&sym)?,
+        Some(ReturnExpr::Direct(e)) => e,
+        None => return None,
+    };
+
+    // Iteratively resolve intermediate variables until no more can be resolved.
+    // Capped at assignments.len() + 1 passes to guard against cyclic assignments.
+    for _ in 0..=assignments.len() {
+        let mut changed = false;
+        for (sym, rhs) in assignments.iter() {
+            let (new_expr, did_change) = expr.clone().substitute_symbol(sym, rhs);
+            if did_change {
+                expr = new_expr;
+                changed = true;
+            }
+        }
+        if !changed {
+            break;
+        }
+    }
+    Some(expr)
+}
+
+enum ReturnExpr {
+    Symbol(InternedString),
+    Direct(Expr),
+}
+
+/// Find the return expression from a function body.
+/// Returns either a symbol identifier (to be resolved via assignments) or
+/// a direct expression (for closures that return complex expressions).
+fn find_return_expr(stmt: &Stmt) -> Option<ReturnExpr> {
+    match stmt.body() {
+        StmtBody::Return(Some(expr)) => {
+            if let ExprValue::Symbol { identifier } = expr.value() {
+                Some(ReturnExpr::Symbol(*identifier))
+            } else {
+                Some(ReturnExpr::Direct(expr.clone()))
+            }
+        }
+        StmtBody::Block(stmts) => {
+            for s in stmts {
+                if let Some(r) = find_return_expr(s) {
+                    return Some(r);
+                }
+            }
+            None
+        }
+        StmtBody::Label { body, .. } => find_return_expr(body),
+        _ => None,
+    }
+}
+
+/// Collect all assignments (symbol = expr) from a statement tree.
+fn collect_assignments(stmt: &Stmt, map: &mut HashMap<InternedString, Expr>) {
+    match stmt.body() {
+        StmtBody::Assign { lhs, rhs } => {
+            if let ExprValue::Symbol { identifier } = lhs.value() {
+                map.insert(*identifier, rhs.clone());
+            }
+        }
+        StmtBody::Decl { lhs, value: Some(val) } => {
+            if let ExprValue::Symbol { identifier } = lhs.value() {
+                map.insert(*identifier, val.clone());
+            }
+        }
+        StmtBody::Block(stmts) => {
+            for s in stmts {
+                collect_assignments(s, map);
+            }
+        }
+        StmtBody::Label { body, .. } => collect_assignments(body, map),
+        _ => {}
+    }
 }
 
 pub fn fn_hooks() -> GotocHooks {

--- a/library/kani_core/src/lib.rs
+++ b/library/kani_core/src/lib.rs
@@ -199,6 +199,17 @@ macro_rules! kani_intrinsics {
 
         #[macro_export]
         macro_rules! forall {
+            // Typed variable: |d: u64 in (lo, hi)| predicate
+            // Uses $t:tt instead of $t:ty because Rust macro rules do not allow
+            // `ty` fragments to be followed by `in`. This means the type must be
+            // a single token (e.g., u64, usize, i32) — path types like
+            // std::num::NonZeroU64 are not supported in this position.
+            (|$i:ident : $t:tt in ($lower_bound:expr, $upper_bound:expr)| $predicate:expr) => {{
+                let lower_bound: $t = $lower_bound;
+                let upper_bound: $t = $upper_bound;
+                let predicate = |$i: $t| $predicate;
+                kani::internal::kani_forall(lower_bound, upper_bound, predicate)
+            }};
             (|$i:ident in ($lower_bound:expr, $upper_bound:expr)| $predicate:expr) => {{
                 let lower_bound: usize = $lower_bound;
                 let upper_bound: usize = $upper_bound;
@@ -213,6 +224,12 @@ macro_rules! kani_intrinsics {
 
         #[macro_export]
         macro_rules! exists {
+            (|$i:ident : $t:tt in ($lower_bound:expr, $upper_bound:expr)| $predicate:expr) => {{
+                let lower_bound: $t = $lower_bound;
+                let upper_bound: $t = $upper_bound;
+                let predicate = |$i: $t| $predicate;
+                kani::internal::kani_exists(lower_bound, upper_bound, predicate)
+            }};
             (|$i:ident in ($lower_bound:expr, $upper_bound:expr)| $predicate:expr) => {{
                 let lower_bound: usize = $lower_bound;
                 let upper_bound: usize = $upper_bound;

--- a/rfc/src/rfcs/0010-quantifiers.md
+++ b/rfc/src/rfcs/0010-quantifiers.md
@@ -199,17 +199,61 @@ The usage of quantifiers should be valid in any part of the Rust code analysed b
 
 Kani should have the same support that CBMC has for quantifiers. For more details, see [Quantifiers](https://github.com/diffblue/cbmc/blob/0a69a64e4481473d62496f9975730d24f194884a/doc/cprover-manual/contracts-quantifiers.md).
 
-The implementation of quantifiers in Kani will be based on the following principle:
+### CBMC constraint: side-effect-free expressions
 
-- **Single expression without function calls**: CBMC's quantifiers only support
-  single expressions without function calls. This means that the CBMC expressions generated
-  from the quantifiers in Kani should be limited to a single expression without any
-  function calls.
+CBMC's quantifiers only support single expressions without function calls or side
+effects. However, even simple Rust expressions like `i + 1` or `i % 2` compile to
+checked arithmetic operations (`OverflowResultPlus`, `checked_rem`) that produce
+`StatementExpression` nodes in the GOTO program — which CBMC rejects as side effects.
 
-To achieve this, we will need to implement the function inlining pass in Kani. This
-  pass will inline all function calls in quantifiers before they are codegened to CBMC
-  expressions. This will ensure that the generated expressions are compliant with CBMC's
-  quantifier support.
+### Pure expression codegen
+
+To solve this, Kani generates **pure expression trees** for quantifier bodies:
+
+1. **Closure body extraction**: `build_quantifier_predicate` in `hooks.rs` extracts
+   the closure's codegen'd body from the symbol table, resolves intermediate variable
+   assignments into a single expression, and substitutes the closure parameter with
+   the quantified variable.
+
+2. **Pure expression inlining**: Before symbol substitution, `inline_as_pure_expr_toplevel`
+   (from `goto_ctx.rs`) flattens all `StatementExpression` nodes by:
+   - Collecting `Decl` assignments from the statement block
+   - Resolving intermediate variables via iterative `substitute_symbol`
+   - Extracting the final expression value
+
+3. **Overflow simplification**: The `Member` handler in the inliner simplifies
+   overflow-checked arithmetic patterns:
+   - `Member(Struct([result, overflowed, padding]), "0")` → extracts `result` directly
+   - `Member(OverflowResultPlus(a, b), "result")` → `Plus(a, b)`
+   - Same for `OverflowResultMinus` → `Minus` and `OverflowResultMult` → `Mult`
+
+   This drops the overflow check, which is acceptable inside quantifier bodies
+   because CBMC evaluates them symbolically.
+
+4. **Function call inlining**: Remaining function calls (e.g., `checked_rem` for `%`)
+   are inlined by looking up the function body in the symbol table, resolving its
+   return expression, and substituting parameters — producing a pure expression.
+
+### Typed quantifier variables
+
+The `forall!` and `exists!` macros support an optional type annotation:
+
+```rust
+kani::forall!(|d: u64 in (1, n)| x % d == 0)
+```
+
+The type is captured as `$t:tt` (not `$t:ty`, since `ty` fragments cannot be
+followed by `in` in macro rules). The internal `kani_forall<T, F>` function is
+already generic over `T`, so the macro simply passes the typed bounds and closure.
+
+### Soundness notes
+
+- **Checked arithmetic dropped**: Overflow checks inside quantifier bodies are
+  silently removed. Division by zero inside quantifier bodies is also not detected.
+  Users must ensure their predicates don't overflow or divide by zero.
+- **Recursive functions**: Recursive function calls in quantifier bodies are detected
+  via a `visited` set and the original expression is returned unchanged (CBMC will
+  reject it as a side effect).
 
 ## Open questions
 

--- a/tests/kani/Quantifiers/arithmetic.rs
+++ b/tests/kani/Quantifiers/arithmetic.rs
@@ -1,0 +1,34 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Z quantifiers
+
+//! Test that arithmetic operations (+, -, *, %) work inside quantifier predicates.
+//! These compile to checked arithmetic (OverflowResultPlus, etc.) which must be
+//! inlined as pure expressions for CBMC to accept them.
+
+#[kani::proof]
+fn check_addition_in_forall() {
+    assert!(kani::forall!(|i in (0, 10)| i + 1 > 0));
+}
+
+#[kani::proof]
+fn check_modulo_in_exists() {
+    assert!(kani::exists!(|i in (0, 10)| i % 2 == 0));
+}
+
+#[kani::proof]
+fn check_multiplication_in_forall() {
+    assert!(kani::forall!(|i in (1, 5)| i * 2 >= 2));
+}
+
+#[kani::proof]
+fn check_subtraction_in_exists() {
+    assert!(kani::exists!(|i in (5, 10)| i - 5 == 0));
+}
+
+#[kani::proof]
+fn check_combined_arithmetic() {
+    let j: usize = kani::any();
+    kani::assume(j % 2 == 0 && j < 2000);
+    kani::assert(kani::exists!(|i in (0, 1000)| i + i == j), "");
+}

--- a/tests/kani/Quantifiers/gcd_invariant.rs
+++ b/tests/kani/Quantifiers/gcd_invariant.rs
@@ -1,0 +1,41 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts -Zloop-contracts -Zquantifiers --solver z3
+
+#![feature(proc_macro_hygiene)]
+#![feature(stmt_expr_attributes)]
+
+/// GCD with a quantifier-based loop invariant proving that the set of common
+/// divisors is preserved across iterations. Uses a typed u64 quantifier
+/// variable with modulo in the predicate body.
+#[kani::requires(x > 0 && y > 0)]
+#[kani::ensures(|&result| result > 0)]
+fn gcd(x: u64, y: u64) -> u64 {
+    let mut a = x;
+    let mut b = y;
+    #[kani::loop_invariant(
+        a > 0
+        // The `d == 0` guard is retained defensively: overflow checks in the
+        // predicate body are dropped during pure expression inlining, so if
+        // `x % d` were evaluated with `d = 0`, it would be undefined behavior.
+        // The range (1, ...) should exclude 0, but the guard provides an extra
+        // layer of safety.
+        && kani::forall!(|d: u64 in (1, a.saturating_add(1))|
+            d == 0 || ((x % d == 0 && y % d == 0)
+            == (a % d == 0 && b % d == 0)))
+    )]
+    while b != 0 {
+        let t = b;
+        b = a % b;
+        a = t;
+    }
+    a
+}
+
+#[kani::proof_for_contract(gcd)]
+#[kani::solver(z3)]
+fn check_gcd() {
+    let x: u64 = kani::any();
+    let y: u64 = kani::any();
+    gcd(x, y);
+}

--- a/tests/kani/Quantifiers/typed_variables.rs
+++ b/tests/kani/Quantifiers/typed_variables.rs
@@ -1,0 +1,17 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Z quantifiers
+
+//! Test that typed quantifier variables work for non-usize integer types.
+
+#[kani::proof]
+#[kani::solver(z3)]
+fn check_typed_u64_forall() {
+    assert!(kani::forall!(|i: u64 in (0, 10)| i < 10));
+}
+
+#[kani::proof]
+#[kani::solver(cvc5)]
+fn check_typed_u64_exists() {
+    assert!(kani::exists!(|i: u64 in (0, 10)| i == 5));
+}

